### ISSUE TITLE
How to set program / instrument in abc string?

### ISSUE THREAD

**Issue by SteveALee** [paulrosen/abcjs#449]
Try as I might I cannot find from the source how to set the instrament / program in the abc string. The docs for `synth.SynthController.setTune` indicates this can be done. And it needs to be done so a click handler can play the correct instrument.

Currently I set the bass instrument (33) in setTune but on click always plays gets instrument 0 until I play the entire tune. Then the voice gets changed on the click hander too

The abs spceification indicates you can use a `%%Midi directive` but thet don;t seem to be implemented.

**Comment by chrisfargen** [comment]
```
I: MIDI=program <n - 1>
```

where `n` = 

```
1 Acoustic Grand
2 Bright Acoustic
3 Electric Grand
...
33 Acoustic Bass
34 Electric Bass (finger)
35 Electric Bass (pick)
...
128 Gunshot
```

**Comment by SteveALee** [comment]
ahah thank you @chrisfargen 

I couldn't find that documented anywhere and was looking at the old %% directives and K or V. Today I learned about inline - I was expecting something in the header.

Works a treat  - thanks again.

I can [see it in the source](https://github.com/paulrosen/abcjs/blob/v5.11.0/src/parse/abc_parse_directive.js#L459) now too!

**Comment by SteveALee** [comment]
> So to get Acoustic Bass (33), it's [I: MIDI=program 32]

did you mean 33 or is somehtign 0 based?

**Comment by chrisfargen** [comment]
> > So to get Acoustic Bass (33), it's [I: MIDI=program 32]
> 
> did you mean 33 or is somehtign 0 based?

Yep, it's zero-based, so you gotta subtract one.

**Comment by SteveALee** [comment]
tah. I seemed odd.

**Comment by ehsansh** [comment]
@SteveALee  @chrisfargen  
I have the same question would you please let me know where is this synth.SynthController.setTune ? 
I want to play the music sheet with different musical instruments sound. Should I do it in abscjs string or it is done in a function?

**Comment by SteveALee** [comment]
@ehsansh  Sure. Originally I used the inline `I` mentioned above in https://musicpracticetools.net but that stopped working when I upgraded to the latest 6.0.0 beta release. I was hoping it would be treated as inline at the start of the body but was not sure what demarks the header. @paulrosen pointed out `K` should be the last item in the header, so I now make sure that is true.

But to answer your question, I use `%% MIDI program nn` in the abc string header to set the intrament for the song. Eg

```abc
X: 1
M: 4/4
Q: 1/4=94
L: 1/8
C: Ahmad Jamal
T: Poinciana
R: Cuban Jazz
%%MIDI program 32
K: C clef=bass transpose=-12
|: EGc :|
``` 
@paulrosen has provided lots of great hosted abc apps, at least one of which can be used to test out strings.

**Comment by ehsansh** [comment]
@SteveALee Wow! That is great. Thanks for the code I did not find it in the documents.
I found a link for instrument soundfonts.
https://gleitz.github.io/midi-js-soundfonts/FluidR3_GM/names.json
Are these the soundfonts that are supported by abcjs? I wish this array has index for each instrument.

I have another question. What is the best way to create soundfont for a musical instrument?

Then I could add soundfont as illustrated here:
https://configurator.abcjs.net/audio
for example:
`abcjs.midi.setSoundFont("mysouondfont.js");`

Am I correct?

**Comment by SteveALee** [comment]
> Thanks for the code I did not find it in the documents.

I understand documentation is a priority for abcjs right now. I runnaged around the code. :)
That code is from my "app" that uses abcjs for my music pratice.

> https://gleitz.github.io/midi-js-soundfonts/FluidR3_GM/names.json
> Are these the soundfonts that are supported by abcjs? I wish this array has index for each instrument.

Yes there are [several available for immediate use](https://github.com/gleitz/midi-js-soundfonts) oooh there are 2 more since I last looked.

> I have another question. What is the best way to create soundfont for a musical instrument?
> Then I could add soundfont as illustrated here:
> https://configurator.abcjs.net/audio
> for example:
> abcjs.midi.setSoundFont("mysouondfont.js");

> Am I correct?

I expect so. I haven't tried building my own sound font *yet*. @paulrosen provides a pointer to building you own on the repository above - sounds like it's a bit fiddly to set up but I'm sure it's just a matter of perseverance :)

Here's how I set the font in 6.0.0Beta12 when [creating the synth](https://github.com/music-practice-tools/music-practice-tools/blob/16d4d98e3d7116f7724093e41b0fd2b461bc554c/src/js/client.js#L425-L439). Note the properties structrue is different in the current release.
